### DISCUSSION TITLE
v1.0

### DISCUSSION THREAD

**Discussion by Adriwin06** [Adriwin06/Ultimate-CommonUI-Menu-System#22] [Releases]
**This is the first version that I would consider "_production ready_"**
This is probably the last release before having to use C++ for things like detecting if the user GPU support Ray Tracing, since this is not exposed in Blueprint. Oh and there is DLSS Ray Reconstruction now if you want ;)

A new and improved CitySample implementation will come soon...

<hr /><em>This discussion was created from the release <a href='https://github.com/Adriwin06/Ultimate-CommonUI-Menu-System/releases/tag/v1.0'>v1.0</a>.</em>

**Comment by Matrix819** [comment]
Needed improvements =) :
Menu - Auto apply checkbox & save settings imidiately -  real time changes
Menu - Auto remember last user position in sub menu - when entering menu should start at precise last changed option (sub menu) - hide menu with ESC. 
Menu - should be narrower (width) (not full screen) to see in real time how picture changes at changing precise option =)
Maybe menu should became only 5 rows height when changing options to have overview over changing video properties in sub menus - rows (LG TVs have this)
Menu shouldn't have darkened right side - to see in real time how picture changes =)
There should be FONT size for menus =) NOW is absolutely to large for 1080p - (slider UI zoom=) like UE editor )
Graphic - GAMMA for brightness is missing at the 1st row =)
Graphic - Scalability is missing at the 2nd row =) - they are the most important
VIDEO - Resolution Scaling / AA should be after Window Mode - the most changing option for testing FSR311, DLSS37, FG
All sub menus should be circular - pressing right - should go from cinematic to low (nvidia reflex off-on-(on+boost)-off-on-(on+boost)-...)

Q,E working for left, right
W,S  NOT working for up / down =)

**Reply by Adriwin06** [reply]
- There is already real time change, just not real time save.
- I will try to make the menu remember the last user focus, but I'm not sure of the way to do it
- You can already see in real time the changes if you click the little eye at the bottom of the right panel, I might remove the color for this part to see better
- I'm not sure to understand what you mean here
- Like I said, I might remove the color in the right panel when you click the little eye
- Why would a font size be needed when the size of the text depend on the size of the window/screen?
- What is missing? You can change the brightness in a sub menu with a slider, what is missing?
- There is scalability settings, why do you say it is missing?
- You mean I should put the Resolution Scaling/Anti-aliasing settings before the vsync?
- Some of them are circular, all of them, I prefer like this, but this is just a little checkbox to tick if you prefer having it like that in your own project
- You can go up or down with page up/down which seem logical to me

**Reply by Matrix819** [reply]
wish list for the MENU to became SUPER USER FRIENDLY =) :
1. real time save - Auto save checkbox - 
2. menu remember the last user focus OR last user change in sub menu =)
3. remove dark colour on the right side of preview window (right side)
4. When user changes option right left > Menu should change to small **COMPACT menu only 5 rows high** - this have LG, SAMSUNG TVs to preview almost all screen for changes in sub menu - THAT IS SO COOL =) And when you clik down next option - your option is still in the middle, just 5 rows changes. THE SIMMILIAR IS GAMMA (OPEN SETTINGS IN NEW WINDOW) and only 1 ROW slider and big preview screen =)
5. we need menu scale option that menu would take smaller part of the 1080p screen (smaller fonts) or menu would have double of current rows ON 4K screen (almost every row could be seen on 4k screeen)
6. Yes, You should put the Resolution Scaling/Anti-aliasing settings before the vsync
7. Yes, I just wanted different orders sub menus at the top row, because they are more used and that's why more important
8. just some menus are circular - If they would be all - that would be better - everything standardized =)
9. Graphic - GAMMA for brightness is missing at the 1st row =) in version 542
10. Graphic - Scalability is missing at the 2nd row =) - they are the most important
11. I was commenting some previous of version 542 and I see 544 menus are better (more important menus are at the top)
12. If you made Q,E working left, right for main menus
13. then make W,S working for up / down sub menus
14. then make A,D working left, right for sub menus change - the same keys we navigete game we should navigate menus - that is logical - isn't it? =) USER FRIENDY =)
15. Disable screen messages to surpress - DLSS information: DLSS not supported AND VIDEO memory exausted expect poor performance (both flashing on my screen very fast) 
16. I checked PS5 dualshock controler - left stick navigating throug all sub menus up down and changing them left right // and L1 and R1 navigating upper 3 menus - that's why we need those keys buttons QWE ASD for menu navigating and changing with only left hand =)

**Reply by Adriwin06** [reply]
Thanks for the suggestions I will add most of them

**Reply by Matrix819** [reply]
Spider man 2 for PS5-Pro has COMPACT menu https://youtu.be/cLQ_fd5JlI8?t=257 
Doom 2016 had Weapons upgrade menu, which remembered position - it was optimisezed that nobody lost time upgrading all weapons to 100% and could made speedrun world record =)

**Reply by Adriwin06** [reply]
> 1. real time save - Auto save checkbox -

Working on it

> 2. menu remember the last user focus OR last user change in sub menu =)

Idk how I will do this

> 3. remove dark colour on the right side of preview window (right side)

Done (last commit)

> 4. When user changes option right left > Menu should change to small **COMPACT menu only 5 rows high** - this have LG, SAMSUNG TVs to preview almost all screen for changes in sub menu - THAT IS SO COOL =) And when you clik down next option - your option is still in the middle, just 5 rows changes. THE SIMMILIAR IS GAMMA (OPEN SETTINGS IN NEW WINDOW) and only 1 ROW slider and big preview screen =)

I still don't understand

> 5. we need menu scale option that menu would take smaller part of the 1080p screen (smaller fonts) or menu would have double of current rows ON 4K screen (almost every row could be seen on 4k screen)

I really don't see what would be the point of doing this since the text scales with the window, so the text is always correctly displayed, and on small screen it is already smaller than on big screens

> 6. Yes, You should put the Resolution Scaling/Anti-aliasing settings before the vsync
> 7. Yes, I just wanted different orders sub menus at the top row, because they are more used and that's why more important

Done (last commit)

> 8. just some menus are circular - If they would be all - that would be better - everything standardized =)

Done (last commit)

> 9. Graphic - GAMMA for brightness is missing at the 1st row =) in version 542

I still don't understand, you mean that it should be at the very top?

> 10. Graphic - Scalability is missing at the 2nd row =) - they are the most important

There is something else that I'm working on that will be at the very top of the graphics settings, you might like it in user friendly things

> 11. I was commenting some previous of version 542 and I see 544 menus are better (more important menus are at the top)

Thanks

> 12. If you made Q,E working left, right for main menus
> 13. then make W,S working for up / down sub menus
> 14. then make A,D working left, right for sub menus change - the same keys we navigete game we should navigate menus - that is logical - isn't it? =) USER FRIENDY =)

I know but Unreal only allows for the keyboards arrows to works for keyboard navigation without deeply changing how it handle input :) and same for the button to validate, which is space

> 15. Disable screen messages to surpress - DLSS information: DLSS not supported AND VIDEO memory exausted expect poor performance (both flashing on my screen very fast)

This is to developer to package their game as a shipping project instead of development build, here it is better to share as development build since it allows you to see errors etc

> 16. I checked PS5 dualshock controler - left stick navigating throug all sub menus up down and changing them left right // and L1 and R1 navigating upper 3 menus - that's why we need those keys buttons QWE ASD for menu navigating and changing with only left hand =)

Like I said above, I can't

**Comment by hollmmes** [comment]
It would be nice to have a video tutorial or documentation showing how we can integrate and migrate it into our own projects.

**Reply by Adriwin06** [reply]
https://github.com/Adriwin06/Ultimate-CommonUI-Menu-System/wiki/Integrate-to-your-own-project

**Reply by hollmmes** [reply]
oh thank you

**Reply by Adriwin06** [reply]
It might be a little bit outdated, but it should still be really useful

**Comment by Matrix819** [comment]
2. maybe the implementation will be complicated and long. If we made a new
copy of the *.ini settings file every time a line was changed, we could
compare which line the change occurred in.
Then when loading the menu during the game, we would immediately open the
menu and give focus to that line
4 If it is possible to make a small transparent compact menu like in
Spiderman 2 https://youtu.be/cLQ_fd5JlI8?t=257 we could put a checkbox use
compact menu - the question is what kind of widgets does unreal common-UI
offer (or does it only allow a full-screen menu? Is it possible to make a
transparent Compact menu that only takes up the bottom left quarter of the
screen?
5 I would like to see smaller menu letters on my big 1080p TV to see more
lines on the screen - is that possible?
9 Yes, I would put Gamma right at the top under GRAPHICS then SCALABILITY
in the 2nd line
GRAPHIC MENU first GAMMA (the demo is so dark gray =)) then SCALABILITY.
BLOOM is also not an option I would like to play with in BENCHMARK - BLOOM
I leave the default =)
I would put colour blind mode under VIDEO at the very bottom (99% of people
have normal vision and don't need it)
VIDEO MENU first RESOLUTION SCALING, THEN DISPLAY (you set the resolution
only once, then forget about it) and finally Colour blind mode at the bottom
12,13,14 Somebody will then need to put a REQUEST or BUG report on UNREAL
to change COMMON-UI to add QWE and ASD for navigation 🤣
15 Considering that you are distributing a Demo for playing, you should*
disable the on-screen messages* when you package it (although I don't have
RTX graphics and I don't have 8GB VRAM, I am not entitled to see 2 messages
flashing at 30 Hz =)
it is especially stupid to see a message that I have too little VRAM (UE
over reacting with VRAM exhausted), even though only 5.5 GB of 6 GB taken
and 500MB of shared RAM are occupied like in Spiderman Remastered and all
games work normally for me - no stuttering .😁
Thanks! 😃

**Reply by Adriwin06** [reply]
> 2. maybe the implementation will be complicated and long. If we made a new copy of the *.ini settings file every time a line was changed, we could compare which line the change occurred in. Then when loading the menu during the game, we would immediately open the menu and give focus to that line 

The .ini files doesn't have anything related to the focus in the menu, that's not what they are made for

> 4. If it is possible to make a small transparent compact menu like in Spiderman 2 https://youtu.be/cLQ_fd5JlI8?t=257 we could put a checkbox use compact menu - the question is what kind of widgets does unreal common-UI offer (or does it only allow a full-screen menu? Is it possible to make a transparent Compact menu that only takes up the bottom left quarter of the screen? 

I don't plan to do that, it is way too much work

> 5. I would like to see smaller menu letters on my big 1080p TV to see more lines on the screen - is that possible? 

Oh you meant scaling the text with the screen size instead of the screen resolution! Idk how I do that

> 9. Yes, I would put Gamma right at the top under GRAPHICS then SCALABILITY in the 2nd line GRAPHIC MENU first GAMMA (the demo is so dark gray =)) then SCALABILITY. 

Well I don't plan to move it in the graphics settings tab, but I'll where to move it in the video settings

> BLOOM is also not an option I would like to play with in BENCHMARK - BLOOM I leave the default =) 

What do you mean? I should remove the option to change the Bloom!?

> I would put color blind mode under VIDEO at the very bottom (99% of people have normal vision and don't need it) VIDEO MENU first RESOLUTION SCALING, THEN DISPLAY (you set the resolution only once, then forget about it) and finally Colour blind mode at the bottom 

Almost like this in the last commit, the updated CitySample is coming soon

> 12,13,14 Somebody will then need to put a REQUEST or BUG report on UNREAL to change COMMON-UI to add QWE and ASD for navigation 🤣 

No it's just up to use to implement it, but it need heavy C++

> 15 Considering that you are distributing a Demo for playing, you should *disable the on-screen messages* when you package it 

I am making a demo to let you change a lot of settings and Cvars, so a dev build is needed for that, and in a dev build there is the on screen messages by default. And seriously, who don't wat to see why the performances are bad by default? If you really don't want it, then just execute the `DisableAllScreenMessages` command (you can since it is a dev build) and it's gone.

> (although I don't have RTX graphics and I don't have 8GB VRAM, I am not entitled to see 2 messages flashing at 30 Hz =) it is especially stupid to see a message that I have too little VRAM (UE over reacting with VRAM exhausted), even though only 5.5 GB of 6 GB taken and 500MB of shared RAM are occupied like in Spiderman Remastered and all games work normally for me - no stuttering .😁 Thanks! 😃

**Reply by Matrix819** [reply]
just delete hires Pictures with small fonts which chrome doesn't want to enlarge anyway
it would be simplier if you just let zoomed-in pictures with large fonts
click to open enlarged picture doesn't work either and pictures are just dubbleing with zoomed in pictures
https://github.com/Adriwin06/Ultimate-CommonUI-Menu-System/wiki/Integrate-to-your-own-project
delete:
Setting up CommonUI in your project - picture at point 4
Setup your Project Settings for these new files  - picture at point 2, 6, 8
Adding Inputs - picture at point 1

btw did you put scalability settings to the top on place 1? in CitySample-5.4_V2.**2**? and uploaded it? =)
https://adriwin78.itch.io/city-sample-ue5-menu/community

**Reply by Adriwin06** [reply]
> just delete hires Pictures with small fonts which chrome doesn't want to enlarge anyway it would be simplier if you just let zoomed-in pictures with large fonts click to open enlarged picture doesn't work either and pictures are just dubbleing with zoomed in pictures https://github.com/Adriwin06/Ultimate-CommonUI-Menu-System/wiki/Integrate-to-your-own-project delete: Setting up CommonUI in your project - picture at point 4 Setup your Project Settings for these new files - picture at point 2, 6, 8 Adding Inputs - picture at point 1
> 

You can just right click -> open in image in new tab

> btw did you put scalability settings to the top on place 1? in CitySample-5.4_V2.**2**? and uploaded it? =) https://adriwin78.itch.io/city-sample-ue5-menu/community

Why 2.**2**? I did not worked on this update yet. And is it still better to have the scalability settings at the top when there is the new Global Quality Preset? (Which is just a placeholder currently but will come)
![image](https://github.com/user-attachments/assets/cb3d57f9-7c87-41c6-aeda-84c558b1a9fa)

**Comment by Matrix819** [comment]
5. Is the option to have scalability for font sizes like UE has under
EDITOR preferences / Application scale? or does common-UI have some
limitations to do it?
9  Bloom should be again at the bottom (rarely used) 😁
15 You should* really really default=**DisableAllScreenMessages *- It is a
working game! Devs can enable them if they want !
Whoever wants to know if there is not enough VRAM - can use OSD Msi
afterburner. It's standard for gamers and devs.
Besides I don't want to see DLSS missing - I know I don't have DLSS 😁
Thanks! 😃

**Comment by hollmmes** [comment]
I really liked your project, but I have some issues. When I add my own map, the map loads, but my character and camera don't move. Also, I think the settings section should be simpler and more minimal. Perhaps having a more straightforward graphics settings menu would be great—it wouldn't make users struggle to understand what the options mean. Can you help with this?

**Reply by hollmmes** [reply]
Also, after loading my level, when I press Tab to open the menu, the mouse doesn't work in the menu.

**Reply by Adriwin06** [reply]
> I really liked your project, 

happy to hear that :)

> but I have some issues. When I add my own map, the map loads, but my character and camera don't move.

It means you didn't implemented it correctly, do you have this in your player controller?
![image](https://github.com/user-attachments/assets/6151c78c-de06-49b7-a521-26ca2c322495)

> Also, I think the settings section should be simpler and more minimal. Perhaps having a more straightforward graphics settings menu would be great—it wouldn't make users struggle to understand what the options mean. Can you help with this?
I do this to provide a wide variety of options to developers, so they can hide/show the ones they prefer :)

**Reply by Adriwin06** [reply]
> Also, after loading my level, when I press Tab to open the menu, the mouse doesn't work in the menu.

It also mean that the menu is not implemented correctly, are you sure you followed the Documentation?

**Reply by hollmmes** [reply]
thank you for help i fixed these, is there a way that we can collapsebutton active by default ? 
![image](https://github.com/user-attachments/assets/e0c26b94-ee8a-4580-a101-c58aff3a369c)

i want to show them like in this picture all collapsed 
![image](https://github.com/user-attachments/assets/2e08f66b-d236-4c58-91da-be785d37457b)

**Reply by Adriwin06** [reply]
Try to set the "start selected" box to true, I can't try right now but it should work.

**Reply by hollmmes** [reply]
I couldn't find it :( When you're available, could you take a screenshot of where it is if it won't take much of your time?

**Reply by Adriwin06** [reply]
Oh sorry it doesn't exist apparently, idk why i was sure it existed

**Reply by Adriwin06** [reply]
But you can do this for each one of them at the event on activated: 
![image](https://github.com/user-attachments/assets/81b4e282-20cb-43d5-983e-7a9db9155a9d)

**Reply by hollmmes** [reply]
thank you boss, do you have any patreon etc. btw

**Reply by Adriwin06** [reply]
I don't have Patreon but I have an itch.io page, a buymeacoffee page and a ko-fi page ☺️

**Reply by hollmmes** [reply]
i'll donate u asap
![image](https://github.com/user-attachments/assets/73ac226b-7010-430b-afb8-8a3bec67e230)
and i tried to set in event on activated but it didnt work 
![image](https://github.com/user-attachments/assets/f8b68d06-b84a-45e5-bc5e-6554c085158d)
Did I do it wrong?

**Reply by Adriwin06** [reply]
I quickly made this ass code and tested it it works. It really isn't what I would call great and dynamic but... it works XD
https://blueprintue.com/blueprint/w8rv-3z-/
it is for the graphic settings, you can adapt it for the other menus as well 

> i'll donate u asap

Thanks you so much 😊
You can access the links in the sponsor part in the repo

**Comment by Prof-Shiba** [comment]
Awesome work!

**Reply by Adriwin06** [reply]
Thanks you really much <3

**Comment by Matrix819** [comment]
never used it before =) just using CTRL+mouse wheel to zoom =)
Nevertheless, pictures are doubled. Better to be just zoomed-in pictures
THEN people reading your instructions will MORE use BRAINS. It will be more
intuitive and easily. =)

*1st scalability settings - very used*
2nd *Visual effects* with many options - rarely used. You Set them up once
- and then you forget about them =)

I don't understand the *global quality placeholder RT Ultra ?? doing what?*

I am afraid to change slider settings if I don't know the default value -*
default value should be in the middle! (more intuitive) - *If user messes
up values THEN user can't change them back to where they were - *users are
lost in the menu =) It's painful. *
It is still painful if you have GLOBAL reset VALUES button, because it
messeas settings users have changed the way they liked - and THEN users are
losing time changing them back to user preference again =)

**Reply by Adriwin06** [reply]
The Global Quality preset does nothing for now, it is placeholder and it written that it is one. And I will inspire myself from the Cyberpunk 2077 menu layout now, it should be good, with quick preset at the top, etc

**Reply by Matrix819** [reply]
UE5.5 is available 12. nov. 2024 - Will you package a new version 5.5 of the demo ? =)
How many hours did you spend designing the menu? I was thinking of starting it myself, but then I saw, oh how wonderful, someone had already done it =)

**Reply by Adriwin06** [reply]
> UE5.5 is available 12. nov. 2024 - Will you package a new version 5.5 of the demo ? =) 

Yes as soon as the required plugins become available for 5.5

> How many hours did you spend designing the menu? 

Idk, a lot XD

> I was thinking of starting it myself, but then I saw, oh how wonderful, someone had already done it =)

Thanks :)

**Reply by Matrix819** [reply]
I'm sure the plugins stayed in the 5.4.4 project, now just open the project in engine 5.5 and it might work.

How many hours does it take to bake into an exe on a 6 core 12600k and SSD and 32GB RAM?

I imported the plugins in a project on 5.4.4, but then I can't open the project in engine 5.5 because it wants VS 2022, which I can't install "windows installer is not available" I installed VS code, which is in the editor list, but 5.5 still doesn't want to open the project because it says I need VS 2022 editor for C++ =)

Engine 5.4.4 asked me if he should use Notepad++ as the editor and I clicked YES - I don't plan on using C++ anyway =)

UE desn't like plugins anyway - FSR311, XESS13, FG cause a strong ghosting effect. Only the built-in upscaler works well and does not cause ghosting when driving a sports car. Maybe AMD and Intel should fix the libraries for FSR11 and XESS ?
Same is in Stalker 2 UE5.1

Even if you don't provide plugins, the demo would work on the DLSS enabler / Uniscaler upscalers - both translate everything to FSR. Just leave the menus for DLSS and FG =)

5.5 has DENOISER - Maybe demo will now look excelent even on low lumen and low shadows setting, whic produced a lot of noise =)

**Reply by Adriwin06** [reply]
> I'm sure the plugins stayed in the 5.4.4 project, now just open the project in engine 5.5 and it might work.

no it's not as simple...

> How many hours does it take to bake into an exe on a 6 core 12600k and SSD and 32GB RAM?

I really don't know, and it depends **a lot** on the project

> I imported the plugins in a project on 5.4.4, but then I can't open the project in engine 5.5 because it wants VS 2022, which I can't install "windows installer is not available" I installed VS code, which is in the editor list, but 5.5 still doesn't want to open the project because it says I need VS 2022 editor for C++ =)

You need VS 2022 to build a .sln solution

> Engine 5.4.4 asked me if he should use Notepad++ as the editor and I clicked YES - I don't plan on using C++ anyway =)

Well you won't be able to build C++ code

> UE desn't like plugins anyway - FSR311, XESS13, FG cause a strong ghosting effect. Only the built-in upscaler works well and does not cause ghosting when driving a sports car. Maybe AMD and Intel should fix the libraries for FSR11 and XESS ? Same is in Stalker 2 UE5.1

Normally it's good, except in Stalker 2

> Even if you don't provide plugins, the demo would work on the DLSS enabler / Uniscaler upscalers - both translate everything to FSR. Just leave the menus for DLSS and FG =)

No, I want to provide quality, so no workaround

> 5.5 has DENOISER - Maybe demo will now look excelent even on low lumen and low shadows setting, whic produced a lot of noise =)

Yeah this is good, I will update the demo with 5.5 when the plugins will be available

**Comment by Matrix819** [comment]
Btw how many hours did you use for the menu design?

Ok leave Global at 1st
Put 2nd scalability settings
3rd visual effects =}

V tor., 19. nov. 2024 23:51 je oseba Adriwin ***@***.***>
napisala:

**Reply by Adriwin06** [reply]
Ok I'll do it

**Comment by Matrix819** [comment]
can't package project city 5.5 - Do you have any idea ? =)
engine 5.5 city 5.5

error1 The windows SDK is not installed properly...Check the SDK section of
the Launch ON menu in the main toolbar to update SDK (I have installed only
12GB game controller picture)
THEN I get PackagingResults: Error: Unknown Error

I am looking this
https://dev.epicgames.com/documentation/en-us/unreal-engine/unreal-engine-5.5-release-notes
my additional install (core dotnet-sdk-3.1.426-win-x64)

   -

   Windows
   -

      Visual Studio 2022 v17.10 or newer (17.12.1 - game controller icon)
      -

      Windows SDK 10.0.22621.0 or newer (10.0.26100.1742 - ALL options)
      -

      LLVM clang
      -

         Minimum: 18.1.3
         -

         Preferred: 18.1.8
         -

      .NET 4.6.2 Targeting Pack (4.8.1)
      -

      .NET 8.0 (9.0 x64 x86)

**Reply by Adriwin06** [reply]
Well this has nothing to do with my project anymore, but the error is pretty clear, you need to install the Windows SDK... or reinstall it.

**Comment by hollmmes** [comment]
how can i setup loading screen, just now it shows up only white/black screen just a 1 seconds then the game starting

**Reply by Adriwin06** [reply]
I use Async Loading Screen, so you need to read its documentation to setup the loading screen

**Reply by Matrix819** [reply]
Who made key P Picture menu?  (only a few settings work)

**Reply by Adriwin06** [reply]
> Who made key P Picture menu?  (only a few settings work)

Wdym "key P Picture menu"? I really don't see what you are talking about.
And which settings are you talking about too?

**Reply by Matrix819** [reply]
Run your CitySample-5.4.4_V2.1 and press keyboard P =)

**Reply by Matrix819** [reply]
do you have private email? not to be all correspondence public?

**Reply by Adriwin06** [reply]
> Run your CitySample-5.4.4_V2.1 and press keyboard P =)

Oh you meant the photo mode... 
Yeah like I said in the release note, it's a little bit broken idk why.

**Reply by Matrix819** [reply]
imaš zasebni e-mail? da ne bo vsa korespondenca javna?

**Comment by Matrix819** [comment]
P Picture Screenshoot menu is again another half baken option from Epic =)
P:\CitySample-5.4.4_V2.1\CitySample\Screenshots

**Comment by hollmmes** [comment]
are you planning to update to 5.5 version ?

**Reply by Adriwin06** [reply]
There is already one: https://github.com/Adriwin06/Ultimate-CommonUI-Menu-System/tree/5.5
I don't merge it to the main since there is not FSR for now
